### PR TITLE
fix:start Celery with a Redis cluster not in the same slot (#4472)

### LINKED ISSUE #4472: Celery with a Redis cluster, an exception is raised: ResponseError("'BRPOP' command keys must be in the same slot")
### Self Checks

- [X] This is only for bug report, if you would like to ask a question, please head to [Discussions](https://github.com/langgenius/dify/discussions/categories/general).
- [X] I have searched for existing issues [search for existing issues](https://github.com/langgenius/dify/issues), including closed ones.
- [X] I confirm that I am using English to submit this report (我已阅读并同意 [Language Policy](https://github.com/langgenius/dify/issues/1542)).
- [X] Please do not modify this template :) and fill in all the required fields.

### Dify version

0.6.8

### Cloud or Self Hosted

Self Hosted (Source)

### Steps to reproduce

1. set CELERY_BROKER_URL=redis://:password@host:port/db
2. celery -A app.celery worker -P gevent -c 1 -Q dataset,generation,mail --loglevel INFO

### ✔️ Expected Behavior

 -------------- celery@TAKATOST.lan v5.2.7 (dawn-chorus)
--- ***** ----- 
-- ******* ---- macOS-10.16-x86_64-i386-64bit 2023-07-31 12:58:08
- *** --- * --- 
- ** ---------- [config]
- ** ---------- .> app:         app:0x7fb568572a10
- ** ---------- .> transport:   redis://:**@localhost:6379/1
- ** ---------- .> results:     postgresql://postgres:**@localhost:5432/dify
- *** --- * --- .> concurrency: 1 (gevent)
-- ******* ---- .> task events: OFF (enable -E to monitor tasks in this worker)
--- ***** ----- 
 -------------- [queues]
                .> dataset          exchange=dataset(direct) key=dataset
                .> generation       exchange=generation(direct) key=generation
                .> mail             exchange=mail(direct) key=mail

[tasks]
  . tasks.add_document_to_index_task.add_document_to_index_task
  . tasks.clean_dataset_task.clean_dataset_task
  . tasks.clean_document_task.clean_document_task
  . tasks.clean_notion_document_task.clean_notion_document_task
  . tasks.create_segment_to_index_task.create_segment_to_index_task
  . tasks.deal_dataset_vector_index_task.deal_dataset_vector_index_task
  . tasks.document_indexing_sync_task.document_indexing_sync_task
  . tasks.document_indexing_task.document_indexing_task
  . tasks.document_indexing_update_task.document_indexing_update_task
  . tasks.enable_segment_to_index_task.enable_segment_to_index_task
  . tasks.generate_conversation_summary_task.generate_conversation_summary_task
  . tasks.mail_invite_member_task.send_invite_member_mail_task
  . tasks.remove_document_from_index_task.remove_document_from_index_task
  . tasks.remove_segment_from_index_task.remove_segment_from_index_task
  . tasks.update_segment_index_task.update_segment_index_task
  . tasks.update_segment_keyword_index_task.update_segment_keyword_index_task

[2023-07-31 12:58:08,831: INFO/MainProcess] Connected to redis://:**@localhost:6379/1
[2023-07-31 12:58:08,840: INFO/MainProcess] mingle: searching for neighbors
[2023-07-31 12:58:09,873: INFO/MainProcess] mingle: all alone
[2023-07-31 12:58:09,886: INFO/MainProcess] pidbox: Connected to redis://:**@localhost:6379/1.
[2023-07-31 12:58:09,890: INFO/MainProcess] celery@TAKATOST.lan ready.

### ❌ Actual Behavior

[2024-05-17 06:19:34,129: INFO/MainProcess] Connected to redis://:**@host:6379/1
[2024-05-17 06:19:34,417: INFO/MainProcess] mingle: searching for neighbors
[2024-05-17 06:19:35,437: CRITICAL/MainProcess] Unrecoverable error: ResponseError("'BRPOP' command keys must in same slot")
Traceback (most recent call last):
  File "/Users/wangvic/miniconda3/lib/python3.11/site-packages/celery/worker/worker.py", line 202, in start
    self.blueprint.start(self)
  File "/Users/wangvic/miniconda3/lib/python3.11/site-packages/celery/bootsteps.py", line 116, in start
    step.start(parent)
  File "/Users/wangvic/miniconda3/lib/python3.11/site-packages/celery/bootsteps.py", line 365, in start
    return self.obj.start()
           ^^^^^^^^^^^^^^^^
  File "/Users/wangvic/miniconda3/lib/python3.11/site-packages/celery/worker/consumer/consumer.py", line 340, in start
    blueprint.start(self)
  File "/Users/wangvic/miniconda3/lib/python3.11/site-packages/celery/bootsteps.py", line 116, in start
    step.start(parent)
  File "/Users/wangvic/miniconda3/lib/python3.11/site-packages/celery/worker/consumer/mingle.py", line 37, in start
    self.sync(c)
  File "/Users/wangvic/miniconda3/lib/python3.11/site-packages/celery/worker/consumer/mingle.py", line 41, in sync
    replies = self.send_hello(c)
              ^^^^^^^^^^^^^^^^^^
  File "/Users/wangvic/miniconda3/lib/python3.11/site-packages/celery/worker/consumer/mingle.py", line 54, in send_hello
    replies = inspect.hello(c.hostname, our_revoked._data) or {}
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/wangvic/miniconda3/lib/python3.11/site-packages/celery/app/control.py", line 389, in hello
    return self._request('hello', from_node=from_node, revoked=revoked)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/wangvic/miniconda3/lib/python3.11/site-packages/celery/app/control.py", line 106, in _request
    return self._prepare(self.app.control.broadcast(
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/wangvic/miniconda3/lib/python3.11/site-packages/celery/app/control.py", line 776, in broadcast
    return self.mailbox(conn)._broadcast(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/wangvic/miniconda3/lib/python3.11/site-packages/kombu/pidbox.py", line 346, in _broadcast
    return self._collect(reply_ticket, limit=limit,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/wangvic/miniconda3/lib/python3.11/site-packages/kombu/pidbox.py", line 388, in _collect
    self.connection.drain_events(timeout=timeout)
  File "/Users/wangvic/miniconda3/lib/python3.11/site-packages/kombu/connection.py", line 341, in drain_events
    return self.transport.drain_events(self.connection, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/wangvic/miniconda3/lib/python3.11/site-packages/kombu/transport/virtual/base.py", line 997, in drain_events
    get(self._deliver, timeout=timeout)
  File "/Users/wangvic/miniconda3/lib/python3.11/site-packages/kombu/transport/redis.py", line 591, in get
    ret = self.handle_event(fileno, event)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/wangvic/miniconda3/lib/python3.11/site-packages/kombu/transport/redis.py", line 573, in handle_event
    return self.on_readable(fileno), self
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/wangvic/miniconda3/lib/python3.11/site-packages/kombu/transport/redis.py", line 569, in on_readable
    chan.handlers[type]()
  File "/Users/wangvic/miniconda3/lib/python3.11/site-packages/kombu/transport/redis.py", line 962, in _brpop_read
    dest__item = self.client.parse_response(self.client.connection,
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/wangvic/miniconda3/lib/python3.11/site-packages/kombu/transport/redis.py", line 247, in parse_response
    ret = super().parse_response(connection, command_name, **options)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/wangvic/miniconda3/lib/python3.11/site-packages/redis/client.py", line 562, in parse_response
    response = connection.read_response()
               ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/wangvic/miniconda3/lib/python3.11/site-packages/redis/connection.py", line 536, in read_response
    raise response
redis.exceptions.ResponseError: 'BRPOP' command keys must in same slot

### PR DESCRIPTION
**Pull request by vicwang3**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes #4472

## Type of Change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update, included: [Dify Document](https://github.com/langgenius/dify-docs)
- [ ] Improvement, including but not limited to code refactoring, performance optimization, and UI/UX improvement
- [ ] Dependency upgrade

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] TODO

# Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [ ] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods
- [ ] `optional` I have made corresponding changes to the documentation 
- [ ] `optional` I have added tests that prove my fix is effective or that my feature works
- [ ] `optional` New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by bowenliang123** [comment]
Some questions for further investigation:
1. how the `global_keyprefix` config of celery helps to fix the slot problem in Redis cluster concepts.
2. What's `{celery_global_prefix}` in the config example? Any reference explanation for the usage ? Is that template holder in some way?

**Comment by binarx2** [comment]
The configuration option should be `result_backend_transport_options` ?

https://docs.celeryq.dev/en/stable/getting-started/backends-and-brokers/redis.html#global-keyprefix

**Comment by vicwang3** [comment]
> Some questions for further investigation:
> 
> 1. how the `global_keyprefix` config of celery helps to fix the slot problem in Redis cluster concepts.
> 2. What's `{celery_global_prefix}` in the config example? Any reference explanation for the usage ? Is that template holder in some way?

The root of the issue is that in Redis cluster mode, keys are distributed across different nodes, and the cluster requires that all keys involved in a specific operation must be located within the same slot since such operations are designed to be executed atomically on a single node.
Solution
Utilizing Redis’s hash tag feature when defining keys for the message queue can ensure that related keys are assigned to the same slot. Users can customize the hash tag CELERY_GLOBAL_PREFIX={celery_global_prefix} in the .env configuration. When operating on REDIS, CELERY_GLOBAL_PREFIX will automatically be used as a prefix. Redis will then assign keys with the same hash tag to the same SLOT.

**Comment by crazywoola** [comment]
This one can be reviewed as well. 
https://github.com/langgenius/dify/pull/4520

File: .gitignore [hunk 1]
```diff
@@ -156,3 +156,4 @@ sdks/python-client/dify_client.egg-info
 .vscode/*
 !.vscode/launch.json
 pyrightconfig.json
+!/api/.env
```
Comment (takatost) [inline .gitignore]: The .env file should be ignored.

File: api/.env.example [hunk 2]
```diff
@@ -22,6 +22,7 @@ FILES_ACCESS_TIMEOUT=300
 
 # celery configuration
 CELERY_BROKER_URL=redis://:difyai123456@localhost:6379/1
+CELERY_GLOBAL_PREFIX={celery_global_prefix}
```
Comment (takatost) [inline api/.env.example]: CELERY_GLOBAL_PREFIX should be left blank by default to avoid affecting existing users

File: api/config.py [hunk 3]
```diff
@@ -209,6 +209,7 @@ def __init__(self):
         self.CELERY_RESULT_BACKEND = 'db+{}'.format(self.SQLALCHEMY_DATABASE_URI) \
             if self.CELERY_BACKEND == 'database' else self.CELERY_BROKER_URL
         self.BROKER_USE_SSL = self.CELERY_BROKER_URL.startswith('rediss://')
+        self.CELERY_GLOBAL_PREFIX = get_env("CELERY_GLOBAL_PREFIX")
```
Comment (takatost) [inline api/config.py]: The default configuration of `global_keyprefix` is an empty string '', please set the default value of CELERY_GLOBAL_PREFIX to an empty string in DEFAULTS of config.py.

File: api/extensions/ext_celery.py [hunk 4]
```diff
@@ -29,6 +29,9 @@ def __call__(self, *args: object, **kwargs: object) -> object:
     celery_app.conf.update(
         result_backend=app.config["CELERY_RESULT_BACKEND"],
         broker_connection_retry_on_startup=True,
+        broker_transport_options={
+            'global_keyprefix': app.config["CELERY_GLOBAL_PREFIX"]
```
Comment (takatost) [inline api/extensions/ext_celery.py]: Carefully consider, suggest setting the `global_keyprefix` key here only when `CELERY_GLOBAL_PREFIX` is not empty.

**Review by takatost — CHANGES_REQUESTED** [review]

**Pull request closed without merging**